### PR TITLE
Modify keras batch normalization to support 2D input

### PR DESCRIPTION
**Pull request by jenniew**
Batch normalization to support 2D input and normalize the second dimension.

Related issue: https://github.com/intel-analytics/analytics-zoo/issues/1222

### CODE REVIEW

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala [hunk 1]
```diff
@@ -55,8 +58,46 @@ class BatchNormalization[T: ClassTag](
    override val gammaInit: String = "one",
    override val dimOrdering: DataFormat = DataFormat.NCHW,
    override val inputShape: Shape = null)(implicit ev: TensorNumeric[T])
-  extends BBatchNormalization[T](
+  extends BKBatchNormalization[T](
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala]: It would be better if you inherit from `KerasLayer` directly as `doBuild` has been overried.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala [hunk 2]
```diff
@@ -55,8 +58,46 @@ class BatchNormalization[T: ClassTag](
...
+    inputShape
+  }
+
+  override def doBuild(inputShape: Shape): AbstractModule[Tensor[T], Tensor[T], T] = {
+    val input = inputShape.toSingle().toArray
+    if(input.length == 4) {
+     super.doBuild(inputShape)
+    } else {
+      val nOutput = input(1)
+      BBatchNormalization[T](nOutput,
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala]: Add an unitest to cover this branch?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala (hunk 1 above)
Comment (jenniew) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala]: Cannot do that because BKBatchNormalization extends KerasLayer[Tensor[T], Tensor[T], T](KerasLayer.addBatch(inputShape)), but KerasLayer.addBatch(inputShape) is a private method

Comment (zhichao-li) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala]: You can use `KerasUtils.addBatch` in this case and `SparseEmbedding.scala` can be a good example.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala (hunk 2 above)
Comment (jenniew) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala]: add test

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala (hunk 1 above)
Comment (jenniew) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala]: updated

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala [hunk 3]
```diff
@@ -49,14 +51,63 @@ import scala.reflect.ClassTag
...
+      // TODO: support arbitrary input shape
+      val layer = SpatialBatchNormalization(
+        nOutput = nChannel,
+        eps = epsilon,
+        momentum = momentum,
+        initWeight = getInit(gammaInit, nChannel),
+        initBias = getInit(betaInit, nChannel),
+        dataFormat = dimOrdering)
+      layer.asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
+    } else {
```
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala]: Maybe you can tighten the condition here as well `input.length == 2`

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala [hunk 4]
```diff
@@ -37,6 +39,32 @@ class BatchNormalizationSpec extends KerasBaseSpec {
...
+    val layer = BatchNormalization[Float](betaInit = "glorot_uniform",
+      gammaInit = "normal", inputShape = Shape(12))
+    seq.add(layer)
+    seq.getOutputShape().toSingle().toArray should be (Array(-1, 12))
+    val input = Tensor[Float](2, 12).rand()
+    val output = seq.forward(input)
+    val gradInput = seq.backward(input, output)
+  }
+
+  "BatchNormalization" should "not work properly for 2D data" in {
```
Comment (zhichao-li) [inline zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala]: this is a 3D input, not 2D

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala [hunk 5]
```diff
@@ -37,6 +39,32 @@ class BatchNormalizationSpec extends KerasBaseSpec {
     val gradInput = seq.backward(input, output)
   }
 
+  "BatchNormalization" should "work properly for 1D data" in {
+    val seq = Sequential[Float]()
+    val layer = BatchNormalization[Float](betaInit = "glorot_uniform",
+      gammaInit = "normal", inputShape = Shape(12))
```
Comment (zhichao-li) [inline zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala]: 2D?

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala [hunk 6]
```diff
@@ -26,7 +28,7 @@ import scala.util.Random
 class BatchNormalizationSpec extends KerasBaseSpec {
 
   // Compared results with Keras on Python side
-  "BatchNormalization" should "work properly" in {
+  "BatchNormalization" should "work properly for 3D data" in {
```
Comment (zhichao-li) [inline zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala]: We can support 2D, 3D and 4D as input as we can see from the unitest?

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala (hunk 4 above)
Comment (jenniew) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala]: updated

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala (hunk 3 above)
Comment (jenniew) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalization.scala]: updated

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala (hunk 5 above)
Comment (jenniew) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala]: updated

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala (hunk 6 above)
Comment (jenniew) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BatchNormalizationSpec.scala]: only 2D and 4D

**Comment by jenniew** [comment]
Jenkins passed: http://172.168.2.101:8080/job/ZOO-PR-Validation-Spark-2.4/460/
